### PR TITLE
Space - Jessica, Nikki, Katie

### PR DESCRIPTION
**Pull request by seaweeddol**
# Assignment Submission: OO Ride Share

Congratulations! You're submitting your assignment. Please reflect on the assignment with these questions.

## Reflection
Question | Answer
:------------- | :-------------
How did getting up to speed on the existing codebase go? If it went well, what worked? If it didn't go well, what will you do differently next time? | It was difficult to figure out what was happening with so many different files. What helped was each of us going through the files on our own, and then coming together with questions and discussing with each other. We would do the same process next time - the scope of the project is what made it difficult, not our process. |
What _inheritance_ relations exist between classes? | Driver, Passenger, and Trips are all children of CsvRecord. |
What _composition_ relations exist between classes? | The TripDispatcher has Drivers, Passengers, and Trips. Each Passenger and Driver has Trips. Each Trip has a Passenger and a Driver. |
Describe a decision you had to make when working on this project. What options were you considering? What helped you make your final decision? | One decision we made was while writing code, we knew that we could write it cleaner, but chose to get it out first and go back and refactor later. This helped us figure out the logic of the code without worrying too much about writing in the most concise way. We also decided to have a 20 minute timer for each person - this was especially useful with a group of three because it allowed everyone to have a chance at being the driver. |
Give an example of a _template method_ that you implemented for this assignment | We implemented the from_csv method in the Driver and Passenger classes. This was a template method in the CsvRecord class. |
Give an example of a _nominal test_ that you wrote for this assignment | We wrote a test to calculate the total revenue of all trips for the Driver class. |
Give an example of an _edge case test_ that you wrote for this assignment | We wrote a test to account for the possibility of a driver having 0 trips, so it would return a total_revenue of 0. |
What is a concept that you gained more clarity on as you worked on this assignment | Relationships between classes (inheritance). Also that organization is very important as a program becomes bigger. Additionally, the importance of having tests, especially as you have more data - it's not as feasible to test with puts statements. |

### CODE REVIEW

File: test/passenger_test.rb [hunk 1]
```diff
@@ -68,7 +69,58 @@
...
+        end_time: Time.parse("2019-01-20 15:48:41 -0800"),
+        cost: 15,
+        rating: 5,
+        driver_id: 5
+        )
+      @passenger.add_trip(trip)
+      @passenger.add_trip(trip2)
+    end
+
+    describe "net_expenditures" do
```
Comment (CheezItMan) [inline test/passenger_test.rb:102]: You should also go back and add tests for incomplete trips.

File: test/driver_test.rb [hunk 2]
```diff
@@ -128,9 +128,85 @@
...
+    end
+
+    it "returns 0 if driver has no trips" do
+      # remove all trips from driver
+      @driver.trips.clear
+
+      expect(@driver.total_revenue).must_equal 0
+    end
+
+    it "returns accurate drivers revenue when in progress trips" do
```
Comment (CheezItMan) [inline test/driver_test.rb:203]: Nice!

**Comment by CheezItMan** [comment]
# OO Ride Share



## Major Learning Goals/Code Review



| Criteria | yes/no, and optionally any details/lines of code to reference |
| --- | --- |
The code demonstrates individual learning about `Time` and the responsibility of `Trip.from_csv`, and uses `Time.parse` in `Trip.from_csv` | ✔️
The code demonstrates breaking out complex logic in helper methods, such as making a helper method in `Trip` to calculate duration | ✔️
There are tests for the nominal cases for the `Passenger#net_expenditures` and `Passenger#total_time_spent` | ✔️
There is at least one edge case test for either `Passenger#net_expenditures` or `Passenger#total_time_spent` testing if the passenger has no trips | ✔️
Practices inheritence. `Driver` inherits from `CsvRecord`, and implements `from_csv` | ✔️
Employs problem-solving and implements `Driver#average_rating` and `Driver#total_revenue` | ✔️
Implements the `TripDispatcher#request_trip`, which creates an instance of `Trip` with a driver and passenger, adds the new trip to `@trips`, and changes the status of the driver | ✔️
Practices composition. In `TripDispatcher#request_trip`, the driver gets connected to the new trip, the passenger gets connected to the new trip | ✔️
Practices git with at least 10 small commits and meaningful commit messages | ✔️, I like how you had good granular commits, you also had good commit messages.

## Testing Requirements

| Testing Requirement | yes/no |
| --- | --- |
There is reasonable test coverage for wave 1, and all wave 1 tests pass | ✔️
There is reasonable test coverage for wave 2, and all wave 2 tests pass | ✔️
Wave 3: Tests in wave 1 and wave 2 explicitly test that only completed trips should be calculated (and ignore in-progress trips) | ✔️
There is reasonable test coverage for `TripDispatcher#request_trip`, and all tests pass | ✔️

## Overall Feedback

| Overall Feedback | Criteria | yes/no |
| --- | --- | --- |
| Green (Meets/Exceeds Standards) | 8+ in Code Review && 3+ in Functional Requirements | ✔️





## Code Style Bonus Awards



Was the code particularly impressive in code style for any of these reasons (or more...?)

| Quality | Yes? |
| --- | --- |
| Perfect Indentation | ✅
| Elegant/Clever | ✅, nice work using enumerables!
| Descriptive/Readable | ✅
| Concise | ✅
| Logical/Organized | ✅





## Summary

Nice work, you hit all the learning goals here.  Great work on getting wave 4 working as well.  Excellent work.

File: lib/trip_dispatcher.rb [hunk 3]
```diff
@@ -27,12 +33,61 @@ def inspect
...
 
+    def pick_driver_for_trip
+      # gets all drivers who are available
+      available_drivers = @drivers.select{ |driver| driver.status == :AVAILABLE}
+      raise "No drivers are currently available" if available_drivers.length == 0
+
+      picked_driver = ""
+      oldest_end_time = Time.now
+
+      available_drivers.each do |driver|
```
Comment (CheezItMan) [inline lib/trip_dispatcher.rb:44]: Awesome!
